src/workflows/recipes/components/FilterObjectForm.js

```diff
@@ -0,0 +1,803 @@
+    // AND IT'S VERY DIFFERENT FROM THE AVAILABLE OPTIONS HERE:
+    // https://normandy.readthedocs.io/en/latest/user/filters.html#filter-context
+    defaultInputOptions: [
+      'normandy.clientId',
+      'normandy.recipe.id',
+      'normandy.request_time',
+      'normandy.userId',
+    ],
+  };
+
+  onSelectSamplingInput = value => {
+    if (!isNaN(parseInt(value, 10))) {
+      value = parseInt(value, 10);
+    }
+    const inputs = this.props.value.input || [];
+    inputs.push(value);
+    inputs.sort((inputA, inputB) => {
+      const smallA = (typeof inputA === 'string' && inputA.toLowerCase()) || inputA.toString();
+      const smallB = (typeof inputB === 'string' && inputB.toLowerCase()) || inputB.toString();
+      if (smallA > smallB) {
+        return 1;
+      } else if (smallA < smallB) {
+        return -1;
+      }
+      return 0;
+    });
+    this.props.bubbleUp({ input: inputs });
+    this.setState({ inputSearch: '', probablyValid: false });
+  };
+
+  onSubmitSamplingInput = () => {
+    // The function rewriteSamplingInput might change the input to make it look like a
+    // string string. I.e. if the input is 'foobar' the output becomes '"foobar"'.
+    this.onSelectSamplingInput(
+      rewriteSamplingInput(this.state.inputSearch.trim(), this.props.defaultInputOptions),
+    );
+  };
+
+  removeInput = value => {
+    const inputs = this.props.value.input.filter(input => input !== value);
+    this.props.bubbleUp({ input: inputs });
+  };
+
+  onSearchInput = value => {
+    // Reason for keeping this as local state is to be able to clear it when something
+    // is chosen.
+    this.setState({ inputSearch: value, probablyValid: !!value.trim() });
+  };
+
+  filterOption = (inputValue, option) => {
+    const regex = makeAutocompleteRegex(inputValue);
+    return regex.test(option.props.children);
+  };
+
+  render() {
+    const { value, disabled } = this.props;
+    const inputs = value.input || [];
+    const inputOptions = this.props.defaultInputOptions.filter(opt => {
+      return !inputs.includes(opt);
+    });
+    return (
+      <FormItem
+        label="Input"
+        name="_sampling.input"
+        required={true}
+        connectToForm={false}
+        {...this.props.formItemLayout}
+      >
+        <AutoComplete
+          dataSource={inputOptions}
+          style={{ width: 200 }}
+          onSelect={this.onSelectSamplingInput}
+          onChange={this.onSearchInput}
+          disabled={disabled}
+          value={this.state.inputSearch}
+          filterOption={this.filterOption}
+          allowClear={true}
+        />
+        <Button
+          type="primary"
+          shape="circle"
+          disabled={disabled || !this.state.probablyValid}
+          icon="plus"
+          onClick={this.onSubmitSamplingInput}
+          // Needed to be able to find button DOM in testing.
+          // https://github.com/kentcdodds/dom-testing-library/issues/108
+          // data-testid="add-button"
```

Comment:
Is this needed or not?

Reply:
Not needed. I went for using CSS selectors even though [Kent recommended against it](https://github.com/kentcdodds/dom-testing-library/issues/108#issuecomment-425502428). What he doesn't know is that Antd is not your grandfathers React component library.